fv3core/stencils/a2b_ord4.py

```diff
@@ -22,485 +36,435 @@
 
 
-@gtstencil()
-def ppm_volume_mean_y(qin: sd, qy: sd):
-    with computation(PARALLEL), interval(...):
-        qy[0, 0, 0] = b2 * (qin[0, -2, 0] + qin[0, 1, 0]) + b1 * (qin[0, -1, 0] + qin)
+@gtscript.function
+def ppm_volume_mean_y(qin: FloatField):
+    return b2 * (qin[0, -2, 0] + qin[0, 1, 0]) + b1 * (qin[0, -1, 0] + qin)
 
 
 @gtscript.function
-def lagrange_y_func(qx):
+def lagrange_y(qx: FloatField):
     return a2 * (qx[0, -2, 0] + qx[0, 1, 0]) + a1 * (qx[0, -1, 0] + qx)
 
 
-@gtstencil()
-def lagrange_interpolation_y(qx: sd, qout: sd):
-    with computation(PARALLEL), interval(...):
-        qout = lagrange_y_func(qx)
-
-
 @gtscript.function
-def lagrange_x_func(qy):
+def lagrange_x(qy: FloatField):
     return a2 * (qy[-2, 0, 0] + qy[1, 0, 0]) + a1 * (qy[-1, 0, 0] + qy)
 
 
-@gtstencil()
-def lagrange_interpolation_x(qy: sd, qout: sd):
-    with computation(PARALLEL), interval(...):
-        qout = lagrange_x_func(qy)
-
-
-@gtstencil()
-def cubic_interpolation_south(qx: sd, qout: sd, qxx: sd):
-    with computation(PARALLEL), interval(...):
-        qxx0 = qxx
-        qxx = c1 * (qx[0, -1, 0] + qx) + c2 * (qout[0, -1, 0] + qxx0[0, 1, 0])
-
+@gtscript.function
+def cubic_interpolation_south(qx: FloatField, qout: FloatField, qxx: FloatField):
+    return c1 * (qx[0, -1, 0] + qx) + c2 * (qout[0, -1, 0] + qxx[0, 1, 0])
 
-@gtstencil()
-def cubic_interpolation_north(qx: sd, qout: sd, qxx: sd):
-    with computation(PARALLEL), interval(...):
-        qxx0 = qxx
-        qxx = c1 * (qx[0, -1, 0] + qx) + c2 * (qout[0, 1, 0] + qxx0[0, -1, 0])
 
+@gtscript.function
+def cubic_interpolation_north(qx: FloatField, qout: FloatField, qxx: FloatField):
+    return c1 * (qx[0, -1, 0] + qx) + c2 * (qout[0, 1, 0] + qxx[0, -1, 0])
 
-@gtstencil()
-def cubic_interpolation_west(qy: sd, qout: sd, qyy: sd):
-    with computation(PARALLEL), interval(...):
-        qyy0 = qyy
-        qyy = c1 * (qy[-1, 0, 0] + qy) + c2 * (qout[-1, 0, 0] + qyy0[1, 0, 0])
 
+@gtscript.function
+def cubic_interpolation_west(qy: FloatField, qout: FloatField, qyy: FloatField):
+    return c1 * (qy[-1, 0, 0] + qy) + c2 * (qout[-1, 0, 0] + qyy[1, 0, 0])
 
-@gtstencil()
-def cubic_interpolation_east(qy: sd, qout: sd, qyy: sd):
-    with computation(PARALLEL), interval(...):
-        qyy0 = qyy
-        qyy = c1 * (qy[-1, 0, 0] + qy) + c2 * (qout[1, 0, 0] + qyy0[-1, 0, 0])
 
+@gtscript.function
+def cubic_interpolation_east(qy: FloatField, qout: FloatField, qyy: FloatField):
+    return c1 * (qy[-1, 0, 0] + qy) + c2 * (qout[1, 0, 0] + qyy[-1, 0, 0])
 
-@gtstencil()
-def qout_avg(qxx: sd, qyy: sd, qout: sd):
-    with computation(PARALLEL), interval(...):
-        qout[0, 0, 0] = 0.5 * (qxx + qyy)
 
+@gtscript.function
+def qout_x_edge(
+    qin: FloatField, dxa: FloatFieldIJ, edge_w: FloatFieldIJ, q2: FloatField
+):
+    """
+    Note:
+        q2 could be 2D, but it is a temporary currently which are all 3D.
```

Comment:
temporaries can be local scalars sometimes too I thought? Does type-hinting force it to be a field? Probably note, the reduction to a scalar probably happens later after inlining I would assume. So from a front-end perspective this note is correct.

Reply:
Yes good point, I'll make that change in the docstring. That was a note to myself and future refactors that because `q2` is accessed with an offset in this function, it can't be a scalar, so it's a 3D field right now, and technically a 2D field is sufficient.

Of course now that I look at it more closely, since we're in a parallel computation, it actually HAS to be 3D. So, I'll remove this comment altogether!